### ISSUE TITLE
'wp create user' doesn't create user with long username, but reports Success

### ISSUE THREAD

**Issue by gitname** [wp-cli/wp-cli#1820]
When I use `wp user create` to create a new user with a *3*-character username, the command output says `Success` (showing a non-zero ID), and subsequent `wp user list` command output shows that user in the table. In this case, my observation matches my expectation.

However, when I use `wp user create` to create a new user with a *80*-character username, the command output says `Success` (this time, showing a zero ID), and subsequent `wp user list` command output does not show that user in the table. In this case, my observation does not match my expectation--I expect either (1) the absence of a `Success` message, or (2) the newly-created user (with the long name) to appear in the user list.

Here is a reproduction of this:

Starting state:
```bash
[vagrant@wordpress wordpress]$ wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
+----+------------+--------------+------------------+---------------------+---------------+
```
Create a user with a short name, then view the user list:
```bash
[vagrant@wordpress wordpress]$ wp user create abc abc@example.com
PHP Notice:  Undefined variable: role in phar:///usr/share/wp-cli/phar/wp-cli.phar/php/commands/user.php on line 309
Notice: Undefined variable: role in phar:///usr/share/wp-cli/phar/wp-cli.phar/php/commands/user.php on line 309
Success: Created user 8.
Password: YbbxRbgRKJNB

[vagrant@wordpress wordpress]$ wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 8  | abc        | abc          | abc@example.com  | 2015-05-06 22:46:05 | subscriber    |
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
+----+------------+--------------+------------------+---------------------+---------------+
```

Create a user with a long name , then view the user list (notice the absence of the newly-created user in the list):
```bash
[vagrant@wordpress wordpress]$ wp user create 12345678vv12345678vv12345678vv12345678vv12345678vv12345678vv12345678vv12345678vv def@example.com
PHP Notice:  Undefined variable: role in phar:///usr/share/wp-cli/phar/wp-cli.phar/php/commands/user.php on line 309
Notice: Undefined variable: role in phar:///usr/share/wp-cli/phar/wp-cli.phar/php/commands/user.php on line 309
Success: Created user 0.
Password: SS67xgQGVuRj

[vagrant@wordpress wordpress]$ wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 8  | abc        | abc          | abc@example.com  | 2015-05-06 22:46:05 | subscriber    |
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
+----+------------+--------------+------------------+---------------------+---------------+
```

Here are details about my environment (this is on the VCCW Vagrant box for WordPress development):

```bash
[vagrant@wordpress wordpress]$ wp --info
PHP binary:     /usr/bin/php
PHP version:    5.4.37
php.ini used:   /etc/php.ini
WP-CLI root dir:        phar://wp-cli.phar
WP-CLI global config:   /home/vagrant/.wp-cli/config.yml
WP-CLI project config:
WP-CLI version: 0.19.0

[vagrant@wordpress wordpress]$ date; hostname; pwd;
Wed May  6 22:45:30 UTC 2015
wordpress.local
/var/www/wordpress
```

**Comment by danielbachhuber** [comment]
> However, when I use wp user create to create a new user with a 80-character username, the command output says Success (this time, showing a zero ID), and subsequent wp user list command output does not show that user in the table.

Which version of WP are you running? On 4.2.2, I get:

```
salty-wordpress ➜  wordpress-test.dev   wp user create 12345678vv12345678vv12345678vv12345678vv12345678vv12345678vv d+test1@danielbachhuber.com
Error: Unknown error creating new user
```

**Comment by gitname** [comment]
I was running WordPress version 4.2.1.
I re-did the same experiment, this time using the command you used, which was slightly different from the one I first used (shorter username and different email address).
I saw the same behavior as I saw before
That behavior is shown here:
```bash
[vagrant@wordpress wordpress]$ wp core version
4.2.1

[vagrant@wordpress wordpress]$ wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 8  | abc        | abc          | abc@example.com  | 2015-05-06 22:46:05 | subscriber    |
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
+----+------------+--------------+------------------+---------------------+---------------+

[vagrant@wordpress wordpress]$ wp user create 12345678vv12345678vv12345678vv12345678vv12345678vv12345678vv d+test1@danielbachhuber.com
PHP Notice:  Undefined variable: role in phar:///usr/share/wp-cli/phar/wp-cli.phar/php/commands/user.php on line 309
Notice: Undefined variable: role in phar:///usr/share/wp-cli/phar/wp-cli.phar/php/commands/user.php on line 309
Success: Created user 0.
Password: !LWK%OMy3gz5

[vagrant@wordpress wordpress]$ wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 8  | abc        | abc          | abc@example.com  | 2015-05-06 22:46:05 | subscriber    |
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
+----+------------+--------------+------------------+---------------------+---------------+
```

I then updated WordPress from 4.2.1 to 4.2.2.
To my surprise, I get the same behavior as I did with WordPress 4.2.1:
```bash
[vagrant@wordpress wordpress]$ wp core version
4.2.1

[vagrant@wordpress wordpress]$ wp core update --version=4.2.2
Updating to version 4.2.2 (en_US)...
Downloading update from https://wordpress.org/wordpress-4.2.2.zip...
Unpacking the update...
Success: WordPress updated successfully.

[vagrant@wordpress wordpress]$ wp core version
4.2.2

[vagrant@wordpress wordpress]$ wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 8  | abc        | abc          | abc@example.com  | 2015-05-06 22:46:05 | subscriber    |
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
+----+------------+--------------+------------------+---------------------+---------------+

[vagrant@wordpress wordpress]$ wp user create 12345678vv12345678vv12345678vv12345678vv12345678vv12345678vv d+test1@danielbachhuber.com
PHP Notice:  Undefined variable: role in phar:///usr/share/wp-cli/phar/wp-cli.phar/php/commands/user.php on line 309
Notice: Undefined variable: role in phar:///usr/share/wp-cli/phar/wp-cli.phar/php/commands/user.php on line 309
Success: Created user 0.
Password: #Xif23UPbgPw

[vagrant@wordpress wordpress]$ wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 8  | abc        | abc          | abc@example.com  | 2015-05-06 22:46:05 | subscriber    |
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
+----+------------+--------------+------------------+---------------------+---------------+
```

**Comment by szepeviktor** [comment]
Try installing wp-cli [by composer](https://github.com/wp-cli/wp-cli/wiki/Alternative-Install-Methods)
and var_dump() [around here](https://github.com/wp-cli/wp-cli/blob/master/php/commands/user.php#L284-L300).

**Comment by gitname** [comment]
I installed wp-cli using the following command:
```bash
[vagrant@wordpress ~]$ composer create-project wp-cli/wp-cli --no-dev
Installing wp-cli/wp-cli (v0.19.0)
  - Installing wp-cli/wp-cli (v0.19.0)
    Downloading: 100%

Created project in /home/vagrant/wp-cli
Loading composer repositories with package information
Installing dependencies
  - Installing wp-cli/php-cli-tools (v0.10.4)
    Downloading: 100%

  - Installing nb/oxymel (v0.1.0)
    Downloading: 100%

  - Installing mustache/mustache (v2.8.0)
    Downloading: 100%

  - Installing ramsey/array_column (1.1.3)
    Downloading: 100%

  - Installing rmccue/requests (v1.6.1)
    Downloading: 100%

  - Installing symfony/finder (v2.6.7)
    Downloading: 100%

Writing lock file
Generating autoload files

[vagrant@wordpress ~]$
```

Then, I ran the experiment again, using this wp-cli instance. The console excerpt below shows this new instance of wp-cli behaves the same way as the old one, when I try to create a user with a 60-character name: 
```bash
[vagrant@wordpress bin]$ /home/vagrant/wp-cli/bin/wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 8  | abc        | abc          | abc@example.com  | 2015-05-06 22:46:05 | subscriber    |
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
| 21 | xyz        | xyz          | xyz@example.com  | 2015-05-11 07:35:20 | subscriber    |
+----+------------+--------------+------------------+---------------------+---------------+

[vagrant@wordpress bin]$ /home/vagrant/wp-cli/bin/wp user create 12345678vv12345678vv12345678vv12345678vv12345678vv12345678vv d+test1@danielbachhuber.com
PHP Notice:  Undefined variable: role in /home/vagrant/wp-cli/php/commands/user.php on line 309
Notice: Undefined variable: role in /home/vagrant/wp-cli/php/commands/user.php on line 309
Success: Created user 0.
Password: p1IRKa4pNAbi

[vagrant@wordpress bin]$ /home/vagrant/wp-cli/bin/wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 8  | abc        | abc          | abc@example.com  | 2015-05-06 22:46:05 | subscriber    |
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
| 21 | xyz        | xyz          | xyz@example.com  | 2015-05-11 07:35:20 | subscriber    |
+----+------------+--------------+------------------+---------------------+---------------+

[vagrant@wordpress bin]$
```
Next, I added `var_dump()` invocations to the file `wp-cli/user.php`, among the lines you highlighted. Here is what that part of my modified `wp-cli/user.php` files looks like:
```php
                if ( is_multisite() ) {
                        $ret = wpmu_validate_user_signup( $user->user_login, $user->user_email );

                     // DEBUG
                        printf("dumping \$ret after calling wpmu_validate_user_signup\n");
                        var_dump($ret);

                        if ( is_wp_error( $ret['errors'] ) && ! empty( $ret['errors']->errors ) ) {
                                WP_CLI::error( $ret['errors'] );
                        }
                        $user_id = wpmu_create_user( $user->user_login, $user->user_email, $user->user_login, $user->user_pass );

                     // DEBUG
                        printf("dumping \$user_id after calling wpmu_create_user\n");
                        var_dump($user_id);

                        if ( ! $user_id ) {
                                WP_CLI::error( "Unknown error creating new user" );
                        }
                        $user->ID = $user_id;
                        $user_id = wp_update_user( $user );

                     // DEBUG
                        printf("dumping \$user_id after calling wp_update_user\n");
                        var_dump($user_id);

                        if ( is_wp_error( $user_id ) ) {
                                WP_CLI::error( $user_id );
                        }
                } else {
                        $user_id = wp_insert_user( $user );

                     // DEBUG
                        printf("dumping \$user_id after calling wp_insert_user\n");
                        var_dump($user_id);

                }

```

With those debug statements in place, I again tried creating a user with a 60-character name. Like before, I saw a success message, but I did not see the new user in the user list.

Here is what the shell prompt showed (now that I have the `var_dump()` invocations in place). According to the debug output, the call to `wp_insert_user(...)` on line 299 of the original file, returned `int(0)`.
```bash
[vagrant@wordpress wp-cli]$ vi /home/vagrant/wp-cli/php/commands/user.php

[vagrant@wordpress wp-cli]$ /home/vagrant/wp-cli/bin/wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 8  | abc        | abc          | abc@example.com  | 2015-05-06 22:46:05 | subscriber    |
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
| 21 | xyz        | xyz          | xyz@example.com  | 2015-05-11 07:35:20 | subscriber    |
+----+------------+--------------+------------------+---------------------+---------------+

[vagrant@wordpress wp-cli]$ /home/vagrant/wp-cli/bin/wp user create 12345678vv12345678vv12345678vv12345678vv12345678vv12345678vv d+test1@danielbachhuber.com
dumping $user_id after calling wp_insert_user
int(0)
PHP Notice:  Undefined variable: role in /home/vagrant/wp-cli/php/commands/user.php on line 329
Notice: Undefined variable: role in /home/vagrant/wp-cli/php/commands/user.php on line 329
Success: Created user 0.
Password: sve#(PyJ*YDH

[vagrant@wordpress wp-cli]$ /home/vagrant/wp-cli/bin/wp user list
+----+------------+--------------+------------------+---------------------+---------------+
| ID | user_login | display_name | user_email       | user_registered     | roles         |
+----+------------+--------------+------------------+---------------------+---------------+
| 8  | abc        | abc          | abc@example.com  | 2015-05-06 22:46:05 | subscriber    |
| 1  | admin      | admin        | vccw@example.com | 2015-05-06 22:12:21 | administrator |
| 21 | xyz        | xyz          | xyz@example.com  | 2015-05-11 07:35:20 | subscriber    |
+----+------------+--------------+------------------+---------------------+---------------+

[vagrant@wordpress wp-cli]$
```
In summary, the new instance of wp-cli behaves the same way the old one did, and the call to wp_insert_user on line 299 returns int(0).

**Comment by szepeviktor** [comment]
Try dumping `$user` before wp_insert_user().

**Comment by danielbachhuber** [comment]
Thanks for the great debugging summary, @gitname.

> In summary, the new instance of wp-cli behaves the same way the old one did, and the call to wp_insert_user on line 299 returns int(0).

Odd. `wp_insert_user()` [isn't _supposed_ to return `0`](https://core.trac.wordpress.org/browser/trunk/src/wp-includes/user.php#L1806) so I suppose the [write operation fails here](https://core.trac.wordpress.org/browser/trunk/src/wp-includes/user.php#L2013).

WordPress core shouldn't blindly assume the write operation was successful. It looks like [this core ticket](https://core.trac.wordpress.org/ticket/32222) might be relevant.
